src/compiler/checker.ts

```diff
@@ -13692,23 +13705,56 @@ namespace ts {
             function createObjectLiteralType() {
                 const stringIndexInfo = isJSObjectLiteral ? jsObjectLiteralIndexInfo : hasComputedStringProperty ? getObjectLiteralIndexInfo(node.properties, offset, propertiesArray, IndexKind.String) : undefined;
                 const numberIndexInfo = hasComputedNumberProperty && !isJSObjectLiteral ? getObjectLiteralIndexInfo(node.properties, offset, propertiesArray, IndexKind.Number) : undefined;
-                const result = createAnonymousType(node.symbol, propertiesTable, emptyArray, emptyArray, stringIndexInfo, numberIndexInfo);
-                const freshObjectLiteralFlag = compilerOptions.suppressExcessPropertyErrors ? 0 : TypeFlags.FreshLiteral;
-                result.flags |= TypeFlags.ContainsObjectLiteral | freshObjectLiteralFlag | (typeFlags & TypeFlags.PropagatingFlags);
-                result.objectFlags |= ObjectFlags.ObjectLiteral;
+                // We used a cached object literal type if the object literal is not the inferred type of a variable
+                // (otherwise Go To Definition on the variable's members wouldn't work properly), has no methods or
+                // accessors (we may cause circularities when obtaining the types of those members), has no index
+                // signatures, is not in a destructuring pattern, doesn't have more than 10 properties, and doesn't
+                // have property names containing colons (which we use as a separator in the cache key).
+                if (!isInferredTypeOfVariable(node) && !hasMethodsOrAccessors && !stringIndexInfo && !numberIndexInfo && !inDestructuringPattern &&
+                    propertiesArray.length <= 10 && every(propertiesArray, propertyNameIsValidKey)) {
+                    return getObjectLiteralType(typeFlags, node.symbol, propertiesTable, propertiesArray);
+                }
+                const result = makeObjectLiteralType(typeFlags, node.symbol, propertiesTable, stringIndexInfo, numberIndexInfo);
                 if (patternWithComputedProperties) {
                     result.objectFlags |= ObjectFlags.ObjectLiteralPatternWithComputedProperties;
                 }
                 if (inDestructuringPattern) {
                     result.pattern = node;
                 }
-                if (!(result.flags & TypeFlags.Nullable)) {
-                    propagatedFlags |= (result.flags & TypeFlags.PropagatingFlags);
-                }
                 return result;
             }
         }
 
+        function isInferredTypeOfVariable(node: Node) {

```

Comment:
what about
```ts
declare function f<T>(t: T): T;
f({ a: 12 }).a
f({ a: 12 }).a/*1*/
```

Won't goto def at `/*1*/` go to the first object literal?

Reply:
No, it will go to the second object literal. But if you then do a goto def on the line above, it also goes to the second object literal. You basically end up going to the first actual instantiation of the type produced by the compiler.

The PR currently shares all object literals types that don't originate from the right hand side of an initializer for a variable with no type annotation. We could exclude other cases (such as arguments to functions), but I don't think the current behavior is totally unreasonable.

Reply:
Here's the worst case that I could come up with. Both `/*1*/` and `/*2*/` will jump to the same definition, I think.

1. I made meaningful names.
2. I split usages across multiple modules.
3. I made two 'different' types that are structurally identical.

(3) is the least realistic here, but this is confusing even if the types are the same. At scale, I think that this is a bug, but maybe not one worth fixing. I guess it boils down to how much excluding arguments hurts performance in our tests?


```ts
// @Filename: functional.ts
export function id<T>(t: T): T { return t; }

// @Filename: people.ts
import { id } from './functional'
export let bob = id({ name: 'Bob', age: 30 });
// @Filename: usage1.ts
import { bob } from './people'
bob.name/*2*/

// @Filename: cars.ts
export let mazda = id({ name: 'cx-5', age: 2 });
// @Filename: usage2.ts
import { mazda } from './cars'

mazda.name/*1*/
```